### PR TITLE
fix: preserve local time when parsing naive datetime strings

### LINKED ISSUE #1277: [BUG]: to_datetime shifts naive strings with omitted or fractional seconds
## Problem

`to_datetime()` interprets equivalent timezone-naive strings differently depending on their precision. Inputs with `HH:mm:ss` retain their wall-clock time in the requested/default timezone, but inputs with `HH:mm` or fractional seconds are interpreted as UTC and then shifted. Near midnight, this silently changes the calendar date.

## Reproduction

On `main` at `940aa1021fdcd10f9fc3a36295a08ca8b21c4155`:

```python
from akkudoktoreos.utils.datetimeutil import to_datetime

for value in (
    "2026-01-15 23:45",
    "2026-01-15 23:45:00",
    "2026-01-15T23:45:00.000",
):
    print(to_datetime(value, in_timezone="Europe/Berlin"))
```

Actual:

```text
2026-01-16 00:45:00+01:00
2026-01-15 23:45:00+01:00
2026-01-16 00:45:00+01:00
```

Expected: all three represent `2026-01-15 23:45:00+01:00`.

The same issue occurs with the `T` separator and with Pendulum's default local timezone set to Europe/Berlin. In summer, the incorrect shift is two hours.

## Affected area and cause

`src/akkudoktoreos/utils/datetimeutil.py`, `to_datetime()`, around lines 924-952.

The explicit format list handles seconds but not omitted/fractional seconds. Remaining strings reach `pendulum.parse(date_input)` without a timezone argument; that parser defaults timezone-naive inputs to UTC. The later timezone conversion shifts their wall-clock time.

The docstring promises local-time interpretation for strings without timezone information. Existing tests establish local interpretation for the supported seconds format. The timezone-conversion example in the docstring should be reconciled with that behavior during this fix.

## Acceptance criteria

- Equivalent naive strings with minutes, seconds, and fractional seconds produce equivalent instants in the requested/default timezone.
- Support both space and `T` date/time separators.
- Add deterministic tests for explicit Europe/Berlin, default local Europe/Berlin, winter and summer dates, and a time near midnight.
- Preserve fractional-second precision.
- Explicit `Z` or numeric UTC offsets retain their instant and continue converting correctly to the target timezone.
- Preserve Unix timestamp behavior and existing supported date formats.
- Keep the implementation focused on this parsing inconsistency and update directly affected documentation examples.
- Run the datetime utility regression tests.

Found while implementing #1273; it is independent and was not fixed as part of that issue.

### PR DESCRIPTION
**Pull request by dr-dimitri**
Fixes #1277.

`to_datetime()` previously interpreted naive strings differently depending on their precision. For example, `2026-01-15 23:45` in Europe/Berlin became `2026-01-16 00:45+01:00`, while the equivalent string with seconds kept the intended date and time. Minutes, seconds, and fractional seconds now consistently represent `2026-01-15 23:45+01:00`.

Pass the resolved target/local timezone to Pendulum's fallback parser. Explicit `Z` and numeric offsets continue to determine the input instant, and Unix timestamps retain their UTC semantics. Update the affected docstring examples to distinguish local wall time from UTC conversion.

Regression coverage includes space and `T` separators, explicit and default Europe/Berlin timezones, winter/summer dates near midnight, fractional-second precision, explicit offsets, and Unix timestamps. The Open-Meteo integration test now independently checks the timestamps and all three irradiance series for all 72 fixture hours, with both UTC and Europe/Berlin as the host timezone. This replaces an expectation that incorrectly associated the 08:00 values with 09:00. The previous parser misplaces all 72 source hours; the corrected parser maps all 72 correctly.

A focused review of the implementation, tests, documentation, and related provider call sites was completed before committing. The patch is limited to three files. Separately identified issues are tracked below.

Validation on Python 3.13.15 with the locked development environment:

- Datetime utility suite: **298 passed**.
- Open-Meteo suite with `--finalize --check-config-side-effect`: **13 passed**.
- Full `make test-ci` on macOS: **1667 passed, 16 skipped, 3 failed, 19 setup errors**. Every datetime and weather test passed. The remaining failures have two independent causes:
  - Two OpenAPI snapshot checks differ only in the commit-derived version string; these checks are skipped on GitHub Actions. Tracked in #1296.
  - One failed server test and 19 server setup errors encounter `psutil.AccessDenied` during system-wide connection inspection in the existing test cleanup. Tracked in #1297.
- Mypy: **passes for all 222 configured files**.
- `make format`: **all hooks pass**, including the locked-environment mypy hook.

The audit also reproduced a similar, pre-existing UTC fallback bug in the separate `to_time` helper. It is tracked in #1295 and is outside this PR's scope.

This pull request was created by an LLM (OpenAI Codex).

### CODE REVIEW

**Comment by b0661** [comment]
Here are my changes - not published yet. 

I  also changed the test:

- tags for better describing the test case.
- explicit wall clock check 

Please regard this:

* fix: to_datetime shifts naive strings with omitted or fractional seconds

```bash
3942d9a4885557b5daafed48f510b5fbf3910e0e
 src/akkudoktoreos/utils/datetimeutil.py |  43 +-
 tests/test_datetimeutil.py              | 736 +++++++++++++++++++++++++++++---
 2 files changed, 703 insertions(+), 76 deletions(-)

diff --git a/src/akkudoktoreos/utils/datetimeutil.py b/src/akkudoktoreos/utils/datetimeutil.py
index a683bf2..80efeb2 100644
--- a/src/akkudoktoreos/utils/datetimeutil.py
+++ b/src/akkudoktoreos/utils/datetimeutil.py
@@ -851,7 +851,14 @@ def to_datetime(
 ) -> Union[DateTime, str]:
     """Convert a date input into a Pendulum DateTime object or a formatted string, with optional timezone handling.
 
-    This function handles various date input formats, adjusts for timezones, and provides flexibility for formatting and time adjustments. For date strings without explicit timezone information, the local timezone is assumed. Be aware that Pendulum DateTime objects created without a timezone default to UTC.
+    This function handles various date input formats, adjusts for timezones,
+    and provides flexibility for formatting and time adjustments. For date
+    strings without explicit timezone information, the date and time are
+    interpreted as wall-clock time in the requested timezone, or in the local
+    timezone when no target timezone is specified. Date strings with an
+    explicit timezone or UTC offset represent an absolute instant and are
+    converted to the requested timezone. Be aware that Pendulum DateTime
+    objects created without a timezone default to UTC.
 
     Args:
         date_input (Optional[Any]): The date input to convert. Supported types include:
@@ -894,7 +901,7 @@ def to_datetime(
         '2024-10-13T00:00:00+00:00'
 
         >>> to_datetime("2024-10-13T15:30:00", in_timezone="Europe/Berlin")
-        DateTime(2024, 10, 13, 17, 30, 0, tzinfo=Timezone('Europe/Berlin'))
+        DateTime(2024, 10, 13, 15, 30, 0, tzinfo=Timezone('Europe/Berlin'))
 
         >>> to_datetime(date(2024, 10, 13), to_maxtime=True)
         DateTime(2024, 10, 13, 23, 59, 59, tzinfo=Timezone('Local'))
@@ -947,14 +954,40 @@ def to_datetime(
                 logger.trace(f"{date_input}, {fmt}, {e}")
                 dt = None
         else:
-            # DateTime input with timezone info
+            # DateTime input with timezone info or an otherwise supported
+            # datetime format.
             try:
                 dt = pendulum.parse(date_input)
+
+                # Pendulum interprets timezone-naive strings as UTC. For
+                # to_datetime(), however, the documented behaviour is that
+                # strings without an explicit timezone represent local
+                # wall-clock time in the requested/default timezone.
+                #
+                # Do not use in_timezone() here: that would convert the
+                # wall-clock value from the implicit UTC timezone. Instead,
+                # attach the target timezone while preserving the date/time
+                # components.
+                has_explicit_timezone = bool(
+                    re.search(
+                        r"(?:Z|[+-]\d{2}:?\d{2})$",
+                        date_input.strip(),
+                        re.IGNORECASE,
+                    )
+                )
+
+                if not has_explicit_timezone:
+                    dt = dt.replace(tzinfo=in_timezone)
+
                 logger.trace(
-                    f"Pendulum Fmt converted: {dt}, tz={dt.tz} from {date_input}, tz={in_timezone}"
+                    f"Pendulum Fmt converted: {dt}, tz={dt.tz} "
+                    f"from {date_input}, tz={in_timezone}"
                 )
             except pendulum.parsing.exceptions.ParserError as e:
-                logger.trace(f"Date string {date_input} does not match any Pendulum formats: {e}")
+                logger.trace(
+                    f"Date string {date_input} does not match any "
+                    f"Pendulum formats: {e}"
+                )
                 dt = None
         if dt is None:
             # Some special values
```

```bash
diff --git a/tests/test_datetimeutil.py b/tests/test_datetimeutil.py
index 9a62943..d9bc4a4 100644
--- a/tests/test_datetimeutil.py
+++ b/tests/test_datetimeutil.py
@@ -841,17 +841,19 @@ class TestPendulumTypes:
 # to_datetime
 # -----------------------------
 
-
-# Test cases for valid pendulum.duration inputs
 @pytest.mark.parametrize(
-    "test_case, local_timezone, date_input, as_string, in_timezone, to_naiv, to_maxtime, expected_output, expected_approximately",
+    (
+    "test_case, local_timezone, date_input, as_string, in_timezone, "
+    "to_naiv, to_maxtime, expected_output, expected_approximately, "
+    "expect_local_wall_time"
+    ),
     [
         # ---------------------------------------
         # from string to pendulum.datetime object
         # ---------------------------------------
-        # - no timezone
+        # - date only / no timezone
         (
-            "TC001",
+            "date_only_utc",
             "Etc/UTC",
             "2024-01-01",
             None,
@@ -860,9 +862,10 @@ class TestPendulumTypes:
             False,
             pendulum.datetime(2024, 1, 1, 0, 0, 0, tz="Etc/UTC"),
             False,
+            False,
         ),
         (
-            "TC002",
+            "date_only_local_berlin",
             "Europe/Berlin",
             "2024-01-01",
             None,
@@ -871,9 +874,10 @@ class TestPendulumTypes:
             False,
             pendulum.datetime(2024, 1, 1, 0, 0, 0, tz="Europe/Berlin"),
             False,
+            False,
         ),
         (
-            "TC003",
+            "date_only_berlin_to_utc",
             "Europe/Berlin",
             "2024-01-01",
             None,
@@ -882,9 +886,10 @@ class TestPendulumTypes:
             False,
             pendulum.datetime(2023, 12, 31, 23, 0, 0, tz="Etc/UTC"),
             False,
+            False,
         ),
         (
-            "TC004",
+            "seconds_space_local_paris",
             "Europe/Paris",
             "2024-01-01 00:00:00",
             None,
@@ -893,9 +898,10 @@ class TestPendulumTypes:
             False,
             pendulum.datetime(2024, 1, 1, 0, 0, 0, tz="Europe/Paris"),
             False,
-        ),
+            False,
+            ),
         (
-            "TC005",
+            "seconds_space_utc_input",
             "Etc/UTC",
             "2024-01-01 00:00:00",
             None,
@@ -904,9 +910,10 @@ class TestPendulumTypes:
             False,
             pendulum.datetime(2024, 1, 1, 1, 0, 0, tz="Europe/Berlin"),
             False,
+            False,
         ),
         (
-            "TC006",
+            "seconds_space_berlin_to_utc",
             "Europe/Berlin",
             "2024-01-01 00:00:00",
             None,
@@ -915,9 +922,10 @@ class TestPendulumTypes:
             False,
             pendulum.datetime(2023, 12, 31, 23, 0, 0, tz="Etc/UTC"),
             False,
+            False,
         ),
         (
-            "TC007",
+            "seconds_space_local_canary",
             "Atlantic/Canary",
             "2024-01-01 12:00:00",
             None,
@@ -934,162 +942,733 @@ class TestPendulumTypes:
                 tz="Atlantic/Canary",
             ),
             False,
+            False,
         ),
         (
-            "TC008",
+            "seconds_space_default_local_timezone",
             "Etc/UTC",
             "2024-01-01 12:00:00",
             None,
-            None,  # force local timezone
+            None,
             None,
             False,
-            pendulum.datetime(2024, 1, 1, 13, 0, 0, tz="Europe/Berlin"),
+            pendulum.datetime(
+                2024,
+                1,
+                1,
+                13,
+                0,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
             False,
         ),
         (
-            "TC009",
+            "seconds_space_berlin_to_utc_explicit_default",
             "Europe/Berlin",
             "2024-01-01 12:00:00",
             None,
             None,
             None,
             False,
-            pendulum.datetime(2024, 1, 1, 11, 0, 0, tz="Etc/UTC"),
+            pendulum.datetime(
+                2024,
+                1,
+                1,
+                11,
+                0,
+                0,
+                tz="Etc/UTC",
+            ),
             False,
+            False,
+        ),
+
+        # ---------------------------------------
+        # Naive strings: explicit Europe/Berlin
+        # ---------------------------------------
+        # These cases specifically protect the regression where
+        # pendulum.parse() interpreted naive values as UTC before
+        # converting them to the requested timezone.
+        #
+        # Winter: CET = UTC+1
+        (
+            "naive_minutes_space_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15 12:34",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
         ),
-        # - with timezone
         (
-            "TC010",
+            "naive_seconds_space_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15 12:34:56",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                56,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_fractional_space_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15 12:34:56.123456",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                56,
+                123456,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_minutes_T_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15T12:34",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_seconds_T_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15T12:34:56",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                56,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_fractional_T_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15T12:34:56.123456",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                56,
+                123456,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+
+        # Summer: CEST = UTC+2
+        (
+            "naive_minutes_space_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15 12:34",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                12,
+                34,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_seconds_space_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15 12:34:56",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                12,
+                34,
+                56,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_fractional_space_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15 12:34:56.123456",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                12,
+                34,
+                56,
+                123456,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_minutes_T_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15T12:34",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                12,
+                34,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_seconds_T_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15T12:34:56",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                12,
+                34,
+                56,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_fractional_T_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15T12:34:56.123456",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                12,
+                34,
+                56,
+                123456,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+
+        # ---------------------------------------
+        # Naive strings: default local timezone
+        # ---------------------------------------
+        (
+            "naive_minutes_default_local_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15 12:34",
+            None,
+            None,
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_seconds_default_local_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15T12:34:56",
+            None,
+            None,
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                12,
+                34,
+                56,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_fractional_default_local_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15T12:34:56.123456",
+            None,
+            None,
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                12,
+                34,
+                56,
+                123456,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+
+        # ---------------------------------------
+        # Naive strings near midnight
+        # ---------------------------------------
+        (
+            "naive_minutes_near_midnight_berlin_winter",
+            "Europe/Berlin",
+            "2024-01-15 23:59",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                15,
+                23,
+                59,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+        (
+            "naive_fractional_near_midnight_berlin_summer",
+            "Europe/Berlin",
+            "2024-07-15T23:59:59.999999",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                15,
+                23,
+                59,
+                59,
+                999999,
+                tz="Europe/Berlin",
+            ),
+            False,
+            True,
+        ),
+
+        # ---------------------------------------
+        # Explicit timezone in string
+        # ---------------------------------------
+        (
+            "slash_date_explicit_berlin",
             "Etc/UTC",
             "02/02/24",
             None,
             "Europe/Berlin",
             None,
             False,
-            pendulum.datetime(2024, 2, 2, 0, 0, 0, tz="Europe/Berlin"),
+            pendulum.datetime(
+                2024,
+                2,
+                2,
+                0,
+                0,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
             False,
         ),
         (
-            "TC011",
+            "offset_datetime_explicit_berlin_winter",
             "Etc/UTC",
-            "2024-03-03T10:20:30.000+01:00",  # No dalight saving time at this date
+            "2024-03-03T10:20:30.000+01:00",
             None,
             "Europe/Berlin",
             None,
-            None,
-            pendulum.datetime(2024, 3, 3, 10, 20, 30, 0, tz="Europe/Berlin"),
+            False,
+            pendulum.datetime(
+                2024,
+                3,
+                3,
+                10,
+                20,
+                30,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
             False,
         ),
         (
-            "TC012",
+            "offset_datetime_explicit_berlin_summer",
             "Etc/UTC",
             "2024-04-04T10:20:30.000+02:00",
             None,
             "Europe/Berlin",
             False,
             None,
-            pendulum.datetime(2024, 4, 4, 10, 20, 30, 0, tz="Europe/Berlin"),
+            pendulum.datetime(
+                2024,
+                4,
+                4,
+                10,
+                20,
+                30,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
             False,
         ),
         (
-            "TC013",
+            "offset_datetime_to_naive",
             "Etc/UTC",
             "2024-05-05T10:20:30.000+02:00",
             None,
             "Europe/Berlin",
             True,
             None,
-            pendulum.naive(2024, 5, 5, 10, 20, 30, 0),
+            pendulum.naive(
+                2024,
+                5,
+                5,
+                10,
+                20,
+                30,
+                0,
+            ),
+            False,
             False,
         ),
-        # - without local timezone as UTC
+
+        # ---------------------------------------
+        # Explicit UTC / offset: instant preserved
+        # ---------------------------------------
         (
-            "TC014",
+            "date_only_explicit_utc",
             "UTC",
             "2024-01-03",
             None,
             "UTC",
             None,
             False,
-            pendulum.datetime(2024, 1, 3, 0, 0, 0, tz="UTC"),
+            pendulum.datetime(
+                2024,
+                1,
+                3,
+                0,
+                0,
+                0,
+                tz="UTC",
+            ),
+            False,
             False,
         ),
         (
-            "TC015",
+            "slash_date_explicit_utc",
             "Atlantic/Canary",
             "02/02/24",
             None,
             "UTC",
             None,
             False,
-            pendulum.datetime(2024, 2, 2, 0, 0, 0, tz="UTC"),
+            pendulum.datetime(
+                2024,
+                2,
+                2,
+                0,
+                0,
+                0,
+                tz="UTC",
+            ),
+            False,
             False,
         ),
         (
-            "TC016",
+            "Z_datetime_is_utc",
             "Atlantic/Canary",
-            "2024-03-03T10:20:30.000Z",  # No dalight saving time at this date
+            "2024-03-03T10:20:30.000Z",
+            None,
+            None,
+            None,
+            None,
+            pendulum.datetime(
+                2024,
+                3,
+                3,
+                10,
+                20,
+                30,
+                0,
+                tz="UTC",
+            ),
+            False,
+            False,
+        ),
+        (
+            "Z_datetime_converts_to_berlin",
+            "Europe/Berlin",
+            "2024-01-15T23:30:00Z",
+            None,
+            "Europe/Berlin",
+            None,
+            False,
+            pendulum.datetime(
+                2024,
+                1,
+                16,
+                0,
+                30,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            False,
+        ),
+        (
+            "offset_datetime_converts_to_berlin",
+            "Europe/Berlin",
+            "2024-07-15T23:30:00+00:00",
             None,
+            "Europe/Berlin",
             None,
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                16,
+                1,
+                30,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
+            False,
+        ),
+        (
+            "offset_fractional_datetime_preserves_precision",
+            "Europe/Berlin",
+            "2024-07-15T23:30:00.123456+00:00",
             None,
+            "Europe/Berlin",
             None,
-            pendulum.datetime(2024, 3, 3, 10, 20, 30, 0, tz="UTC"),
+            False,
+            pendulum.datetime(
+                2024,
+                7,
+                16,
+                1,
+                30,
+                0,
+                123456,
+                tz="Europe/Berlin",
+            ),
+            False,
             False,
         ),
+
         # ---------------------------------------
         # from pendulum.datetime to pendulum.datetime object
         # ---------------------------------------
         (
-            "TC017",
+            "pendulum_naive_datetime_uses_utc",
             "Atlantic/Canary",
             pendulum.datetime(2024, 4, 4, 0, 0, 0),
             None,
             None,
             None,
             False,
-            pendulum.datetime(2024, 4, 4, 0, 0, 0, tz="Etc/UTC"),
+            pendulum.datetime(
+                2024,
+                4,
+                4,
+                0,
+                0,
+                0,
+                tz="Etc/UTC",
+            ),
+            False,
             False,
         ),
         (
-            "TC018",
+            "pendulum_naive_datetime_converts_to_berlin",
             "Atlantic/Canary",
             pendulum.datetime(2024, 4, 4, 1, 0, 0),
             None,
             "Europe/Berlin",
             None,
             False,
-            pendulum.datetime(2024, 4, 4, 3, 0, 0, tz="Europe/Berlin"),
+            pendulum.datetime(
+                2024,
+                4,
+                4,
+                3,
+                0,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
             False,
         ),
         (
-            "TC019",
+            "pendulum_utc_datetime_converts_to_berlin",
             "Atlantic/Canary",
-            pendulum.datetime(2024, 4, 4, 1, 0, 0, tz="Etc/UTC"),
+            pendulum.datetime(
+                2024,
+                4,
+                4,
+                1,
+                0,
+                0,
+                tz="Etc/UTC",
+            ),
             None,
             "Europe/Berlin",
             None,
             False,
-            pendulum.datetime(2024, 4, 4, 3, 0, 0, tz="Europe/Berlin"),
+            pendulum.datetime(
+                2024,
+                4,
+                4,
+                3,
+                0,
+                0,
+                tz="Europe/Berlin",
+            ),
+            False,
             False,
         ),
         (
-            "TC020",
+            "pendulum_berlin_datetime_converts_to_utc",
             "Atlantic/Canary",
-            pendulum.datetime(2024, 4, 4, 2, 0, 0, tz="Europe/Berlin"),
+            pendulum.datetime(
+                2024,
+                4,
+                4,
+                2,
+                0,
+                0,
+                tz="Europe/Berlin",
+            ),
             None,
             "Etc/UTC",
             None,
             False,
-            pendulum.datetime(2024, 4, 4, 0, 0, 0, tz="Etc/UTC"),
+            pendulum.datetime(
+                2024,
+                4,
+                4,
+                0,
+                0,
+                0,
+                tz="Etc/UTC",
+            ),
+            False,
             False,
         ),
+
         # ---------------------------------------
         # from string to UTC string
         # ---------------------------------------
         # - no timezone
         #   local timezone UTC
         (
-            "TC021",
+            "naive_seconds_utc_string_from_utc",
             "Etc/UTC",
             "2023-11-06T00:00:00",
             "UTC",
@@ -1098,10 +1677,11 @@ class TestPendulumTypes:
             None,
             "2023-11-06T00:00:00Z",
             False,
+            False,
         ),
-        #    local timezone "Europe/Berlin"
+        #   local timezone Europe/Berlin
         (
-            "TC022",
+            "naive_seconds_utc_string_from_berlin",
             "Europe/Berlin",
             "2023-11-06T00:00:00",
             "UTC",
@@ -1110,10 +1690,14 @@ class TestPendulumTypes:
             None,
             "2023-11-05T23:00:00Z",
             False,
+            False,
         ),
-        # - no microseconds
+
+        # ---------------------------------------
+        # UTC string conversion with explicit offset
+        # ---------------------------------------
         (
-            "TC023",
+            "offset_seconds_utc_string",
             "Atlantic/Canary",
             "2024-10-30T00:00:00+01:00",
             "UTC",
@@ -1122,9 +1706,10 @@ class TestPendulumTypes:
             None,
             "2024-10-29T23:00:00Z",
             False,
+            False,
         ),
         (
-            "TC024",
+            "offset_seconds_lowercase_utc_string",
             "Atlantic/Canary",
             "2024-10-30T01:00:00+01:00",
             "utc",
@@ -1133,10 +1718,10 @@ class TestPendulumTypes:
             None,
             "2024-10-30T00:00:00Z",
             False,
+            False,
         ),
-        # - with microseconds
         (
-            "TC025",
+            "offset_fractional_utc_string",
             "Atlantic/Canary",
             "2024-10-07T10:20:30.000+02:00",
             "UTC",
@@ -1145,14 +1730,14 @@ class TestPendulumTypes:
             None,
             "2024-10-07T08:20:30Z",
             False,
+            False,
         ),
+
         # ---------------------------------------
         # from None to pendulum.datetime object
         # ---------------------------------------
-        # - no timezone
-        #   local timezone
         (
-            "TC026",
+            "none_returns_current_datetime",
             None,
             None,
             None,
@@ -1161,6 +1746,7 @@ class TestPendulumTypes:
             None,
             pendulum.now(),
             True,
+            False,
         ),
     ],
 )
@@ -1175,9 +1761,11 @@ def test_to_datetime(
     to_maxtime,
     expected_output,
     expected_approximately,
+    expect_local_wall_time,
 ):
     """Test pendulum.datetime conversion with valid inputs."""
     set_other_timezone(local_timezone)
+
     result = to_datetime(
         date_input,
         as_string=as_string,
@@ -1185,32 +1773,38 @@ def test_to_datetime(
         to_naiv=to_naiv,
         to_maxtime=to_maxtime,
     )
-    # if isinstance(date_input, str):
-    #    print(f"Input:    {date_input}")
-    # else:
-    #    print(f"Input:    {date_input} tz={date_input.timezone}")
+
     if isinstance(expected_output, str):
-        # print(f"Expected: {expected_output}")
-        # print(f"Result:   {result}")
         assert result == expected_output
     elif expected_output.timezone is None:
-        # We expect an exception
+        # We expect an exception.
         with pytest.raises(TypeError):
             assert compare_datetimes(result, expected_output).equal
     else:
         compare = compare_datetimes(result, expected_output)
-        # print(f"---- Testcase:  {test_case} ----")
-        # print(f"Expected: {expected_output} tz={expected_output.timezone}")
-        # print(f"Result:   {result} tz={result.timezone}")
-        # print(f"Compare:  {compare}")
+
         if expected_approximately:
-            # Compare using Unix timestamps to avoid pendulum timezone state
-            # issues and to prevent staleness from parametrize collection time.
+            # Compare using Unix timestamps to avoid Pendulum timezone
+            # state issues and to prevent staleness from parametrization
+            # collection time.
             import time
 
             assert abs(result.timestamp() - time.time()) < 300
         else:
-            assert compare.equal == True
+            assert compare.equal
+
+        if expect_local_wall_time:
+            # A naive input string must be interpreted as a wall-clock time
+            # in the requested/default timezone, rather than being parsed as
+            # UTC first and subsequently converted.
+            assert result.timezone.name == expected_output.timezone.name
+            assert result.year == expected_output.year
+            assert result.month == expected_output.month
+            assert result.day == expected_output.day
+            assert result.hour == expected_output.hour
+            assert result.minute == expected_output.minute
+            assert result.second == expected_output.second
+            assert result.microsecond == expected_output.microsecond
 
 
 # -----------------------------
```

File: src/akkudoktoreos/utils/datetimeutil.py [hunk 1]
```diff
@@ -974,9 +983,10 @@ def to_datetime(
                 logger.trace(f"{date_input}, {fmt}, {e}")
                 dt = None
         else:
-            # DateTime input with timezone info
+            # The fallback also handles naive strings with omitted or fractional seconds.
+            # Supply their assumed timezone; explicit offsets take precedence in Pendulum.
             try:
-                dt = cast(pendulum.DateTime, pendulum.parse(date_input))
+                dt = cast(pendulum.DateTime, pendulum.parse(date_input, tz=timezone))
```
Comment (b0661) [inline src/akkudoktoreos/utils/datetimeutil.py:995]: Please see the solution given in the comments.

**Comment by dr-dimitri** [comment]
@b0661 Thanks for sharing your proposal. The clearer test case names and explicit wall-clock checks are useful suggestions. We agree on the intended behavior: naive strings should keep their wall-clock time in the requested/default timezone, while explicit offsets should preserve the input instant.

My recommendation would be to keep the current `pendulum.parse(date_input, tz=timezone)` implementation and selectively incorporate the test improvements. In the pinned Pendulum 3.2.0 implementation, a parsed explicit offset already takes precedence over the supplied `tz`.

I checked a concrete edge case in the proposed regex-and-replace approach: Pendulum accepts the short offset `+02`, but the regex does not recognize it. For `2026-01-15T12:34+02`, with Europe/Berlin as the target timezone:

- The current PR correctly returns `2026-01-15T11:34:00+01:00`.
- The proposed logic, even using the resolved Berlin timezone, returns `2026-01-15T12:34:00+01:00`, changing the instant by one hour.

The new PR tests already compare ISO representations, including the date, wall time, fractional seconds, and offset. Explicit timezone-name assertions for the naive-string cases and a regression case for short offsets would be useful additions. Clearer test IDs could also be incorporated without restructuring the whole existing test table.

I reran the current datetime utility suite: **298 passed**. No code changes have been made in response to the review yet.

Would you prefer to keep the current parser fix with those targeted test improvements, or take another approach? Please decide how you would like us to proceed.

**Comment by b0661** [comment]
> Would you prefer to keep the current parser fix with those targeted test improvements, or take another approach? 

I am fine with the current parser fix. I would like to have the test improvements to be used. The explanation about on how the wall clock is preserved for the configured location in the case of native datetimes should be made more explicit in the current parser fix.

**Comment by dr-dimitri** [comment]
@b0661 The requested changes have been implemented and pushed in 627e8c0.

- Kept the existing `pendulum.parse(date_input, tz=timezone)` parser fix.
- Expanded the docstring and inline explanation of how naive strings preserve the configured location's wall-clock time, including a concrete Europe/Berlin example.
- Added descriptive test case names and explicit assertions for the timezone, date, and clock components.
- Added coverage for short and compact UTC offsets, including `+02`.

Validation: **383 datetime and Open-Meteo tests passed**, the full mypy check passed, and formatting/lint checks passed.

The changes are ready for your review.
